File: go/samples/signature/signature.go
package main

import (
	"github.com/bloock/bloock-sdk-go/v2"
	"github.com/bloock/bloock-sdk-go/v2/builder"
	"github.com/bloock/bloock-sdk-go/v2/client"
	"github.com/bloock/bloock-sdk-go/v2/client/entity"
	"github.com/bloock/bloock-sdk-quickstart/utils"
	"github.com/fatih/color"
)

func main() {
	utils.Sample("signature", func(c utils.Config) error {
		bloock.ApiKey = c.ApiKey
		sdk := client.NewClient()

		keys, err := sdk.GenerateKeys()
		if err != nil {
			return err
		}

		signedRecord, err := builder.NewRecordBuilderFromString("Hello world").
			WithSigner(entity.NewEcsdaSigner(keys.PrivateKey)).
			Build()

		if err != nil {
			return err
		}

		color.Green("[✓] Record was signed successfully")

		// we can add another signature with a different key
		keys, err = sdk.GenerateKeys()
		if err != nil {
			return err
		}

        color.Yellow("[+] Adding another signature")
		signedRecord, err = builder.NewRecordBuilderFromRecord(signedRecord).
			WithSigner(entity.NewEcsdaSigner(keys.PrivateKey)).
			Build()

		if err != nil {
			return err
		}

		color.Green("[✓] Record was signed successfully")

		hash, err := signedRecord.GetHash()
		if err != nil {
			return err
		}
		color.Green("[✓] Hash: %s", hash)

		signatures, err := signedRecord.GetSignatures()
		if err != nil {
			return err
		}

		for i, signature := range signatures {
			color.Green("[✓] Signature %d: %+v", i+1, signature.Signature)
		}

		return nil
	})
}

File: java/src/main/java/com/bloock/quickstart/samples/signature/Signature.java
package com.bloock.quickstart.samples.signature;

import com.bloock.quickstart.utils.Config;
import com.bloock.quickstart.utils.Logger;
import com.bloock.quickstart.utils.Sample;
import com.bloock.sdk.Bloock;
import com.bloock.sdk.builder.Builder;
import com.bloock.sdk.client.Client;
import com.bloock.sdk.entity.EcsdaSigner;
import com.bloock.sdk.entity.Keys;
import com.bloock.sdk.entity.Record;

public class Signature extends Sample {
  public void run(Config config) throws Exception {
    Bloock.apiKey = config.apiKey;
    Client client = new Client();

    Keys keys = client.generateKeys();

    Record signedRecord =
        Builder.fromString("Hello world").withSigner(new EcsdaSigner(keys.getPrivateKey())).build();

    Logger.success("Record was signed successfully");

    // we can add another signature with a different key
    Logger.info("Adding another signature");
    keys = client.generateKeys();

    signedRecord =
        Builder.fromRecord(signedRecord).withSigner(new EcsdaSigner(keys.getPrivateKey())).build();

    Logger.success("Record was signed successfully");

    Logger.success("Hash: " + signedRecord.getHash());

    signedRecord.getSignatures().stream()
        .forEach(
            signature -> {
              Logger.success("Signature " + signature.getSignature());
            });
  }

  Signature(String name) {
    super(name);
  }

  public static void main(String[] args) {
    new Signature("Signature");
  }
}
